packages/state-transition/src/block/processDeposit.ts

```diff
@@ -41,68 +49,132 @@ export function processDeposit(fork: ForkSeq, state: CachedBeaconStateAllForks,
   // deposits must be processed in order
   state.eth1DepositIndex += 1;
 
-  const pubkey = deposit.data.pubkey; // Drop tree
-  const amount = deposit.data.amount;
-  const cachedIndex = epochCtx.pubkey2index.get(pubkey);
+  applyDeposit(fork, state, deposit.data);
+}
+
+/**
+ * Adds a new validator into the registry. Or increase balance if already exist.
+ * Follows applyDeposit() in consensus spec. Will be used by processDeposit() and processDepositReceipt()
+ *
+ */
+export function applyDeposit(
+  fork: ForkSeq,
+  state: CachedBeaconStateAllForks,
+  deposit: DepositData | DepositReceipt
+): void {
+  const {config, validators, epochCtx} = state;
+  const {pubkey, withdrawalCredentials, amount} = deposit;
+
+  const cachedIndex = epochCtx.getValidatorIndex(pubkey);
   if (cachedIndex === undefined || !Number.isSafeInteger(cachedIndex) || cachedIndex >= validators.length) {
-    // verify the deposit signature (proof of posession) which is not checked by the deposit contract
-    const depositMessage = {
-      pubkey: deposit.data.pubkey, // Retain tree for hashing
-      withdrawalCredentials: deposit.data.withdrawalCredentials, // Retain tree for hashing
-      amount: deposit.data.amount,
-    };
-    // fork-agnostic domain since deposits are valid across forks
-    const domain = computeDomain(DOMAIN_DEPOSIT, config.GENESIS_FORK_VERSION, ZERO_HASH);
-    const signingRoot = computeSigningRoot(ssz.phase0.DepositMessage, depositMessage, domain);
-    try {
-      // Pubkeys must be checked for group + inf. This must be done only once when the validator deposit is processed
-      const publicKey = bls.PublicKey.fromBytes(pubkey, CoordType.affine, true);
-      const signature = bls.Signature.fromBytes(deposit.data.signature, CoordType.affine, true);
-      if (!signature.verify(publicKey, signingRoot)) {
-        return;
+    if (isValidDepositSignature(config, pubkey, withdrawalCredentials, amount, deposit.signature)) {
+      addValidatorToRegistry(fork, state, pubkey, withdrawalCredentials, amount);
+    }
+  } else {
+    if (fork < ForkSeq.electra) {
+      // increase balance by deposit amount right away pre-electra
+      increaseBalance(state, cachedIndex, amount);
+    } else if (fork >= ForkSeq.electra) {
+      const stateElectra = state as CachedBeaconStateElectra;
+      const pendingBalanceDeposit = ssz.electra.PendingBalanceDeposit.toViewDU({
+        index: cachedIndex,
+        amount: BigInt(amount),
+      });
+      stateElectra.pendingBalanceDeposits.push(pendingBalanceDeposit);
```

Comment:
so we add new `PendingBalanceDeposit` with `amount` but inside `switchToCompoundingValidator()` we also add another excess `PendingBalanceDeposit` with `balance - MIN_ACTIVATION_BALANCE`

is it correct to add 2 `PendingBalanceDeposit` for a single deposit? it's not an issue of our implementation just a question for the spec

Reply:
seems to be correct, so balance - MIN_ACTIVATION_BALANCE is the excess which would have been partially withdrawan, but is not queued to add to balance deposits